x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java

```diff
@@ -103,47 +109,43 @@ public void clusterChanged(ClusterChangedEvent event) {
-                    // no previous knowledge of existing nodes. Consequently, if a model was manually removed (task-kill) from a node
-                    // it may get re-allocated to that node when another node is added/removed...
-
-                    // As this produces a cluster state update task, we are certain that if the persistent
-                    // task framework results in assigning some ML tasks on that same cluster state change
-                    // we do not end up over-allocating a node. Both this service and the persistant task service
-                    // will produce a cluster state update but the one that gets applied first wins. The other
-                    // update will be rejected and we will retry to assign getting a correct update on available memory
-                    // on each node.
-                    // Also, note that as this service is a returned as a component of the ML plugin,
-                    // and components are created before persistent task executors, we will try to allocate
-                    // trained models before we try to assign ML persistent tasks.
-                    return addRemoveAssignmentNodes(currentState);
-                }
-
-                @Override
-                public void onFailure(Exception e) {
-                    logger.warn("failed to allocate models", e);
-                }
+        if (clusterService.state().nodes().getMinNodeVersion().before(DISTRIBUTED_MODEL_ALLOCATION_VERSION)) {
+            // we should not try to rebalance assignments while there may be nodes running on a version
+            // prior to introducing distributed model allocation.
+            return;
+        }
 
-                @Override
-                public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
-                    logger.trace(
+        if (event.localNodeMaster() && shouldAllocateModels(event)) {
+            // TODO this has a weird side-effect for allocating to nodes
+            // If the event indicates there were nodes added/removed, this method only looks at the current state and has
+            // no previous knowledge of existing nodes. Consequently, if a model was manually removed (task-kill) from a node
+            // it may get re-allocated to that node when another node is added/removed...
+            //
+            // As this produces a cluster state update task, we are certain that if the persistent
+            // task framework results in assigning some ML tasks on that same cluster state change
+            // we do not end up over-allocating a node. Both this service and the persistent task service
+            // will produce a cluster state update but the one that gets applied first wins. The other
+            // update will be rejected and we will retry to assign getting a correct update on available memory
+            // on each node.
+            rebalanceAssignments(
+                clusterService.state(),
```

Comment:
why are we using `clusterService.state()` here and not the ones provided in the event?

Reply:
FYI, 

```
        state.set(newClusterState);

        callClusterStateListeners(clusterChangedEvent, stopWatch);
```
It seems to be that the `clusterService.state` is set and then all the listeners called. But I am not sure I would rely on that and use the `event` state wherever possible to prevent accidental race conditions.

Reply:
+++